clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/vmimage/view/AzureVMImageLookupController.groovy

```diff
@@ -0,0 +1,189 @@
+        creds.vmImages.each { vmImage ->
+          def imageName = vmImage.offer + "-" + vmImage.sku + " (Recommended)"
+          if (lookupOptions.q == null || imageName.toLowerCase().startsWith(lookupOptions.q.toLowerCase())) {
+            result += new AzureNamedImage(
+              imageName: imageName,
+              publisher: vmImage.publisher,
+              offer: vmImage.offer,
+              sku: vmImage.sku,
+              version: vmImage.version
+            )
+            if (result.size() >= MAX_SEARCH_RESULTS)
+              return result
+          }
+        }
+      }
+    }
+
+    if (!lookupOptions.configOnly && lookupOptions.q != null && lookupOptions.q.length() >= MIN_NAME_FILTER) {
+      // TODO: retrieve the list of virtual machine images from the azure respective cache
+      result.addAll(getAllMatchingKeyPattern(lookupOptions.q,
+                              Keys.getVMImageKey( azureCloudProvider,
+                                                  lookupOptions.account?:'*',
+                                                  lookupOptions.region?:'*',
+                                                  "*",
+                                                  "*"
+                              )))
+      //result.addAll(cachedVMImages.subList(0, Math.min(MAX_SEARCH_RESULTS, cachedVMImages.size())))
+    }
+
+    result
+  }
+
+  List<AzureNamedImage> getAllMatchingKeyPattern(String vmImagePartName, String pattern) {
+    loadResults(vmImagePartName, cacheView.filterIdentifiers(Keys.Namespace.AZURE_VMIMAGES.ns, pattern))
+  }
+
+  List<AzureNamedImage> loadResults(String vmImagePartName, Collection<String> identifiers) {
+    def results = [] as List<AzureNamedImage>
+    def data = cacheView.getAll(Keys.Namespace.AZURE_VMIMAGES.ns, identifiers, RelationshipCacheFilter.none())
+    data.each {cacheData ->
+      def item = fromCacheData(vmImagePartName, cacheData)
+
+      if (item)
+        results += item
+
+      if (results.size() >= MAX_SEARCH_RESULTS)
+        return results
+    }
+
+    results
+  }
+
+  AzureNamedImage fromCacheData(String vmImagePartName, CacheData cacheData) {
+    AzureVMImage vmImage = objectMapper.convertValue(cacheData.attributes['vmimage'], AzureVMImage)
+    def imageName = vmImage.offer + "-" + vmImage.sku + " (${vmImage.publisher} ${vmImage.version})"
+    def parts = Keys.parse(azureCloudProvider, cacheData.id)
+
+    if (imageName.toLowerCase().startsWith(vmImagePartName.toLowerCase())) {
+
+      return new AzureNamedImage(
+                  imageName: imageName,
+                  publisher: vmImage.publisher,
+                  offer: vmImage.offer,
+                  sku: vmImage.sku,
+                  version: vmImage.version,
+                  account: parts.account,
+                  region: parts.region
+                )
+    }
+
+    null
+  }
+
+  @ResponseStatus(value = HttpStatus.NOT_FOUND, reason = 'Image not found')
```

Comment:
Minor detail but you probably don't want to include a `reason` as it will override any specific message you provide when throwing the exception.